clang/lib/Frontend/CompileJobCacheResult.cpp

```diff
@@ -34,6 +34,39 @@ Error CompileJobCacheResult::forEachOutput(
   return Error::success();
 }
 
+Error CompileJobCacheResult::forEachLoadedOutput(
```

Comment:
Does this matter as we are moving towards build system integration? You can get a lot more parallelism from the driver. Async APIs are good, but it should not be on the `CompileJobCacheResult` which is quite narrow.

Also swift compiler is using a completely different model that `CompileJobCacheResult` always contains a singular output and I secretly want to move clang towards the direction of one cache lookup per output file if that is not expensive (cache lookup should be async too)

Reply:
> Does this matter as we are moving towards build system integration? You can get a lot more parallelism from the driver.

It matters because we'll be supporting CMake/Make projects that don't have build system integration.

> Async APIs are good, but it should not be on the `CompileJobCacheResult` which is quite narrow.

`CompileJobCacheResult::forEachLoadedOutput` is a convenience function for "wait to get all outputs, then walk them", the function itself is synchronous for the caller. `CompileJobCacheResult` doesn't offer async APIs but the caller can get the output ID it wants and load it asynchronously, see the "FIXME: We wait to materialize each module file before proceeding" in this PR for an example in code where it would be beneficial to do that.

> Also swift compiler is using a completely different model that `CompileJobCacheResult` always contains a singular output and I secretly want to move clang towards the direction of one cache lookup per output file if that is not expensive (cache lookup should be async too)

Are you suggesting having separate cache lookups for each kind of output (like separate cache lookup for diagnostics)? What is the benefit compared to bundling the related outputs together in one cache lookup?

Reply:
Right, for cmake support, we can keep it.

We can discuss the benefit of separate cache key offline unless we want to switch now. The benefit is smaller for clang, mainly:
* support cache hit if the command-line generates less output (remember we don't support all the outputs yet, like remarks, stats, other random things)
* we can flatten the CompileJobCacheResult to be a direct lookup, thus the lookup speed should be the same if CAS/Cache speed is comparable (but compute all the keys are slightly more expensive)
* swift needs replay/lookup clang module via mechanism in clang so the result schema needs to be in sync.

Reply:
> we can flatten the CompileJobCacheResult to be a direct lookup, thus the lookup speed should be the same if CAS/Cache speed is comparable (but compute all the keys are slightly more expensive)

The thing to be mindful of is not local computation, but the number of network trips. Each cache lookup adds an extra network cache lookup that blocks initiating the network CAS load for the output, so they introduce latency.
Even if we have separate cache lookups for additional outputs, we should at least bundle outputs that are always generated together, like diagnostics.

We can discuss more offline.